synapse/appservice/__init__.py

```diff
@@ -164,9 +164,9 @@ async def _matches_user(
         does_match = await self.matches_user_in_member_list(event.room_id, store)
         return does_match
 
-    @cached(num_args=1)
+    @cached(num_args=1, cache_context=True)
     async def matches_user_in_member_list(
-        self, room_id: str, store: "DataStore"
+        self, room_id: str, store: "DataStore", cache_context,
```

Comment:
Mypy [doesn't seem to understand](https://buildkite.com/matrix-dot-org/synapse/builds/13074#a5063c17-acbd-4c52-a016-526f00246981) `on_invalidate=cache_context.invalidate` so I've left it untyped. There might be a way to make this work, but I couldn't find any examples in the Synapse where this is done.

Reply:
This doesn't seem to have to do with the typing on the input here, but the call to `get_users_in_room`. mypy is still failing.

Reply:
I guess not. I'm basically struggling to understand why https://github.com/matrix-org/synapse/blob/develop/synapse/handlers/presence.py#L1126-L1128 is fine but my call is not?

Reply:
Adding `on_invalidate=None` to the function args broke several tests by changing the output (how!?). I've instead used the horrid `**kwargs`, which seems to make the tests happy again.

Reply:
This should be fixed by: https://github.com/matrix-org/synapse/pull/8684

The problem is that we do :sparkles: magic :mage_man: :sparkles: rewriting of the function signatures for `@cached` functions, and we forgot o add the `on_invalidate` function to it.

> I guess not. I'm basically struggling to understand why https://github.com/matrix-org/synapse/blob/develop/synapse/handlers/presence.py#L1126-L1128 is fine but my call is not?

It's because we hadn't typed the `hs` arg, and so mypy didn't know the function signature of `get_users_in_room` for that case.

Reply:
Thank you for fixing it :heart:. I noticed there was magic, but would not have come up with #8684 by myself

Reply:
I've removed **kwargs.